unsoundness (race condition)

You are implementing `Record` for `SyncRecord<'a>`, which contains a `fmt::Arguments<'a>`. This allows triggering a race condition (https://github.com/rust-lang/rust/issues/45197), that will be fixed in Rust in the future (https://github.com/rust-lang/rust/pull/45198).

I checked what happens when only implementing `Record` for `SyncRecord<'static>`, but that makes `woodpecker` unusable:

```
error[E0597]: borrowed value does not live long enough
   --> src/logger.rs:578:22
    |
578 |                 log!(">>{}<<", "unconditional");
    |                 -----^^^^^^^^-------------------
    |                 |    |
    |                 |    does not live long enough
    |                 in this macro invocation
    |
   ::: src/macros/helpers.rs
    |
33  |         })
    |         - temporary value only lives until here
    |
    = note: borrowed value must be valid for the static lifetime...

error[E0597]: borrowed value does not live long enough
   --> src/logger.rs:733:35
    |
733 |                             info!("{}", idx);
    |                             ------^^^^-------
    |                             |     |
    |                             |     does not live long enough
    |                             in this macro invocation
    |
   ::: src/macros/helpers.rs
    |
33  |         })
    |         - temporary value only lives until here
    |
    = note: borrowed value must be valid for the static lifetime...

error[E0597]: `idx` does not live long enough
   --> src/logger.rs:733:41
    |
733 |                             info!("{}", idx);
    |                                         ^^^ does not live long enough
    |
   ::: src/macros/helpers.rs
    |
33  |         })
    |         - borrowed value only lives until here
    |
    = note: borrowed value must be valid for the static lifetime...

error[E0597]: borrowed value does not live long enough
   --> src/logger.rs:735:32
    |
735 |                         debug!("{}", idx);
    |                         -------^^^^-------
    |                         |      |
    |                         |      does not live long enough
    |                         in this macro invocation
    |
   ::: src/macros/helpers.rs
    |
33  |         })
    |         - temporary value only lives until here
    |
    = note: borrowed value must be valid for the static lifetime...

error[E0597]: `idx` does not live long enough
   --> src/logger.rs:735:38
    |
735 |                         debug!("{}", idx);
    |                                      ^^^ does not live long enough
    |
   ::: src/macros/helpers.rs
    |
33  |         })
    |         - borrowed value only lives until here
    |
    = note: borrowed value must be valid for the static lifetime...

error[E0597]: borrowed value does not live long enough
   --> src/logger.rs:733:35
    |
733 |                             info!("{}", idx);
    |                             ------^^^^-------
    |                             |     |
    |                             |     does not live long enough
    |                             in this macro invocation
    |
   ::: src/macros/logging.rs
    |
86  |                     root.log(&RECORD, format_args!($($arg)*));
    |                                                              - temporary value only lives until here
    |
    = note: borrowed value must be valid for the static lifetime...
    = note: consider using a `let` binding to increase its lifetime

error[E0597]: `idx` does not live long enough
   --> src/logger.rs:733:41
    |
733 |                             info!("{}", idx);
    |                                         ^^^ does not live long enough
    |
   ::: src/macros/logging.rs
    |
88  |             });
    |             - borrowed value only lives until here
    |
    = note: borrowed value must be valid for the static lifetime...

error[E0597]: borrowed value does not live long enough
   --> src/logger.rs:735:32
    |
735 |                         debug!("{}", idx);
    |                         -------^^^^-------
    |                         |      |
    |                         |      does not live long enough
    |                         in this macro invocation
    |
   ::: src/macros/logging.rs
    |
86  |                     root.log(&RECORD, format_args!($($arg)*));
    |                                                              - temporary value only lives until here
    |
    = note: borrowed value must be valid for the static lifetime...
    = note: consider using a `let` binding to increase its lifetime

error[E0597]: `idx` does not live long enough
   --> src/logger.rs:735:38
    |
735 |                         debug!("{}", idx);
    |                                      ^^^ does not live long enough
    |
   ::: src/macros/logging.rs
    |
88  |             });
    |             - borrowed value only lives until here
    |
    = note: borrowed value must be valid for the static lifetime...
```

I'm not sure how to proceed here.

--- a/src/record/imp.rs
+++ b/src/record/imp.rs
@@ -25,6 +25,8 @@ use std::sync::Arc;
 use std::fmt;
 use std::fmt::Write;
 
+use std::marker::PhantomData;
+
 use formatters::Formatter;
 use levels::LogLevel;
 use record::Record;
@@ -143,6 +145,7 @@ pub(crate) struct SyncRecord<'a> {
     args: fmt::Arguments<'a>,
     precord: RecordLazyMeta,
     ts: time::Timespec,
+    _oibit_remover: PhantomData<*mut Fn()>,
 }
 
 impl<'a> SyncRecord<'a> {
@@ -156,6 +159,7 @@ impl<'a> SyncRecord<'a> {
             ts: ts,
             args: args,
             precord: RecordLazyMeta::new(formatter),
+            _oibit_remover: PhantomData,
         }
     }
 }

--- a/src/record/mod.rs
+++ b/src/record/mod.rs
@@ -23,7 +23,7 @@ use levels::LogLevel;
 
 /// Log record that holds information where log was recorded
 /// and the message details.
-pub trait Record: Send + Sync {
+pub trait Record {
     /// Log level of the record.
     fn level(&self) -> LogLevel;
 

